### PR TITLE
feat: add onSuccess and onError callbacks

### LINKED ISSUE #31: Add back `onSuccess` and `onError` handlers to `Form`
I actually didn't realize these were gone until I tried to use them 😆 

So, from reading the docs, I figure the rationale was that the onSuccess/onError callbacks were used to update state to show some message in the UI. The `useFormSubmit` state works fine in that case. But when I need to fire some side effect, the hook state feels more cumbersome.

Here's an example and comparison. The modal accepts a render prop for a `close` function.

Without the `onSuccess` callback, I need to use a `useEffect` that relies on the hook state, which also requires making a new component in which to use the hook:
```js
export function EditBucketButton({ bucket }) {
  return (
    <Modal
      title="edit bucket"
      renderContent={({ close }) => (
        <EditBucketForm bucket={bucket} onSuccess={close} />
      )}
    />
  )
}

function EditBucketForm({ bucket, onSuccess }) {
  const { isSuccess } = useFormSubmit("edit-bucket")

  useEffect(() => {
    if (isSuccess) onSuccess()
  }, [isSuccess, onSuccess])

  return (
    <Form name="edit-bucket" action={`/buckets/${bucket.id}`} method="patch">
      {/* form stuff */}
    </Form>
  )
}
```

With the callback, I don't need a new component. I can write the form inline, and pass the close callback to `onSuccess`:
```js
export function EditBucketButton({ bucket }) {
  return (
    <Modal
      title="edit bucket"
      renderContent={({ close }) => (
        <Form name="edit-bucket" action={`/buckets/${bucket.id}`} method="patch" onSuccess={close}>
          {/* form stuff */}
        </Form>
      )}
    />
  )
}
```
Summary: the callbacks seem valuable as a convenient way to run side effects in response to a form submission.

In this specific case, I could refactor `Modal` to just accept props, then I could have all of the state + the effect in one component, but writing an effect for each instance of this still doesn't sound like fun 😅 That, and [there are other APIs that work like this](https://headlessui.dev/react/disclosure#closing-disclosures-manually), as a point of reference.

### PR DESCRIPTION
**Pull request by itsMapleLeaf**
Closes #31

### CODE REVIEW

File: src/form.tsx [hunk 1]
```diff
@@ -181,16 +196,49 @@ type FormState<Data extends Record<string, unknown> = Record<string, unknown>> =
...
+  const onErrorRef = useLatestRef(onError);
+  useEffect(() => {
+    if (!initializedRef.current) return;
+    if (state.status === 'error') {
+      onErrorRef.current?.(state);
+    }
+  }, [onErrorRef, state]);
+
+  useEffect(() => {
+    // reset form state when we mount
```
Comment (smeijer) [inline src/form.tsx:231]: I'm not sure if we should do that. This means someone can't conditionally show the form without losing state. Think, not showing the firm during loading state, and reverting to open form on error.

Comment (smeijer) [reply src/form.tsx:231]: Maybe we need to add a `reset` method to FormState?

Comment (itsMapleLeaf) [reply src/form.tsx:231]: Yeah, I initially thought this was needed to ensure the form doesn't immediately call the error/success callback on mount, but that's what the ref is for. I'll remove the state reset

**Comment by smeijer** [comment]
Thanks! Much appreciated.

**Comment by smeijer** [comment]
Released with https://github.com/smeijer/next-runtime/releases/tag/v2.1.0

**Pull request merged**